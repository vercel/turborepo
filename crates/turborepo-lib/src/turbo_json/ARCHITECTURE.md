# Turborepo Architecture: Configuration and Task Definition Loading

This document explains the process of loading `turbo.json` files and creating task definitions in Turborepo.
The process involves several phases that transform raw configuration files into an executable task graph.
The loading and resolving of task definitions is driven during task graph construction.

## Overview

The configuration and task loading process follows this high-level flow:

1. **Configuration Resolution**: Collect and merge configuration from multiple sources
2. **TurboJson Loading**: Parse `turbo.json` files into raw structures
3. **Task Processing**: Convert raw definitions to processed intermediate representation with DSL token handling
4. **Task Definition Resolution**: Transform processed definitions into final validated structures
5. **Task Graph Construction**: Build the executable task graph from resolved definitions

## Phase 1: Configuration Resolution

### Sources and Priority

Configuration is collected from multiple sources with the following priority (highest to lowest):

1. Command line arguments
2. Environment variables
3. Override environment variables
4. Local configuration (`.turbo/config.json`)
5. Global authentication (`~/.turbo/auth.json`)
6. Global configuration (`~/.turbo/config.json`)
7. Turbo.json configuration

### Key Components

- **`TurborepoConfigBuilder`** (`crates/turborepo-lib/src/config/mod.rs`): Orchestrates the configuration loading process
- **`TurboJsonReader`** (`crates/turborepo-lib/src/config/turbo_json.rs`): Extracts configuration options from the root `turbo.json` file
- **`ConfigurationOptions`**: The final merged configuration structure

## Phase 2: TurboJson Loading

### Key Components

- **`TurboJsonLoader`** (`crates/turborepo-lib/src/turbo_json/loader.rs`): Loads and resolves turbo.json files
- **`RawTurboJson`**: Raw deserialized structure from JSON files
- **`TurboJson`**: Validated structure containing raw task definitions

### Process

1. **File Discovery**: Locate `turbo.json` or `turbo.jsonc` files
2. **Parsing**: Deserialize JSON into `RawTurboJson` structures
3. **Basic Validation**: Convert to `TurboJson` with structural validation
4. **Workspace Resolution**: Apply workspace-specific overrides

## Phase 3: Processed Task Definition (Intermediate Representation)

### Key Components

- **`ProcessedTaskDefinition`** (`crates/turborepo-lib/src/turbo_json/processed.rs`): Intermediate representation with DSL token processing
- **`ProcessedGlob`**: Parsed glob patterns with separated components (base pattern, negation flag, turbo_root flag)
- **`ProcessedInputs`/`ProcessedOutputs`**: Collections of processed globs with resolution methods

### Processing Steps

1. **DSL Token Detection**: Identify and separate `$TURBO_ROOT$` and `!` prefixes from glob patterns
2. **Early Validation**: Validate absolute paths and invalid token usage at parse time with span information
3. **Prefix Stripping**: Store clean glob patterns without DSL prefixes
4. **Component Separation**: Track negation and turbo_root requirements as separate boolean flags

## Phase 4: Task Definition Resolution

### Key Components

- **`RawTaskDefinition`**: Raw task configuration from JSON
- **`ProcessedTaskDefinition`**: Intermediate representation with parsed DSL tokens
- **`TaskDefinition`**: Final validated and resolved task configuration
- **`TaskId`** and **`TaskName`** (from `turborepo-task-id` crate): Task identification types

### Transformation Process

The resolution now follows a three-stage pipeline:

1. **Raw → Processed** (`ProcessedTaskDefinition::from_raw`):

   - Parse glob patterns and extract DSL tokens
   - Validate absolute paths with span information
   - Strip prefixes and store components separately

2. **Processed → Resolved** (`TaskDefinition::from_processed`):

   - Apply `$TURBO_ROOT$` token replacement using `resolve()` methods
   - Parse `dependsOn` into topological and task dependencies
   - Extract environment variables from `env` and `passThroughEnv`
   - Transform outputs into inclusion/exclusion patterns
   - Validate configuration consistency (e.g., interactive tasks can't be cached)

3. **Inheritance**: The `extend` module handles merging multiple `ProcessedTaskDefinition`s before final resolution
