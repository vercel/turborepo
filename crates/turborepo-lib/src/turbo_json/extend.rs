//! Module for code related to "extends" behavior for task definitions

use super::processed::{
    ProcessedDependsOn, ProcessedEnv, ProcessedInputs, ProcessedOutputs, ProcessedPassThroughEnv,
    ProcessedTaskDefinition, ProcessedWith,
};

/// Trait for types that can be merged with extends behavior
trait Extendable {
    /// Merges another instance into self.
    /// If the other instance has `extends: true`, it extends the current value.
    /// Otherwise, it replaces the current value.
    fn extend(&mut self, other: Self);
}

/// Macro to handle the extend/replace logic for a field
macro_rules! merge_field_vec {
    ($self:ident, $other:ident, $field:ident) => {
        if $other.extends {
            $self.$field.extend($other.$field);
        } else {
            $self.$field = $other.$field;
        }
    };
}

impl Extendable for ProcessedDependsOn {
    fn extend(&mut self, other: Self) {
        merge_field_vec!(self, other, deps);
        self.extends = other.extends;
    }
}

impl Extendable for ProcessedEnv {
    fn extend(&mut self, other: Self) {
        merge_field_vec!(self, other, vars);
        // Sort and dedup for env vars
        if other.extends {
            self.vars.sort();
            self.vars.dedup();
        }
        self.extends = other.extends;
    }
}

impl Extendable for ProcessedOutputs {
    fn extend(&mut self, other: Self) {
        merge_field_vec!(self, other, globs);
        self.extends = other.extends;
    }
}

impl Extendable for ProcessedPassThroughEnv {
    fn extend(&mut self, other: Self) {
        merge_field_vec!(self, other, vars);
        // Sort and dedup for env vars
        if other.extends {
            self.vars.sort();
            self.vars.dedup();
        }
        self.extends = other.extends;
    }
}

impl Extendable for ProcessedWith {
    fn extend(&mut self, other: Self) {
        merge_field_vec!(self, other, tasks);
        self.extends = other.extends;
    }
}

impl Extendable for ProcessedInputs {
    fn extend(&mut self, other: Self) {
        merge_field_vec!(self, other, globs);
        // Handle the default flag specially
        if other.extends {
            // When extending, OR the default flags
            self.default = self.default || other.default;
        } else {
            // When replacing, use the other's default
            self.default = other.default;
        }
        self.extends = other.extends;
    }
}

impl FromIterator<ProcessedTaskDefinition> for ProcessedTaskDefinition {
    fn from_iter<T: IntoIterator<Item = ProcessedTaskDefinition>>(iter: T) -> Self {
        iter.into_iter()
            .fold(ProcessedTaskDefinition::default(), |mut def, other| {
                def.merge(other);
                def
            })
    }
}

macro_rules! set_field {
    ($this:ident, $other:ident, $field:ident) => {{
        if let Some(field) = $other.$field {
            $this.$field = field.into();
        }
    }};
}

macro_rules! merge_field {
    ($this:ident, $other:ident, $field:ident) => {{
        if let Some(other_field) = $other.$field {
            match &mut $this.$field {
                Some(self_field) => {
                    // Merge using the Mergeable trait
                    self_field.extend(other_field);
                }
                None => {
                    // No existing value, just set it
                    $this.$field = Some(other_field);
                }
            }
        }
    }};
}

impl ProcessedTaskDefinition {
    // Merges another ProcessedTaskDefinition into this one
    // Array fields use the Mergeable trait to handle extends behavior
    pub fn merge(&mut self, other: ProcessedTaskDefinition) {
        // Array fields that support extends behavior
        merge_field!(self, other, outputs);
        merge_field!(self, other, depends_on);
        merge_field!(self, other, inputs);
        merge_field!(self, other, env);
        merge_field!(self, other, pass_through_env);
        merge_field!(self, other, with);

        // Non-array fields that are simply replaced
        let other_has_range = other.cache.as_ref().is_some_and(|c| c.range.is_some());
        let self_does_not_have_range = self.cache.as_ref().is_some_and(|c| c.range.is_none());

        if other.cache.is_some()
            // If other has range info and we're missing it, carry it over
            || (other_has_range && self_does_not_have_range)
        {
            self.cache = other.cache;
        }
        set_field!(self, other, output_logs);
        set_field!(self, other, persistent);
        set_field!(self, other, interruptible);
        set_field!(self, other, interactive);
        set_field!(self, other, env_mode);
    }
}

#[cfg(test)]
mod test {
    use turborepo_errors::Spanned;
    use turborepo_unescape::UnescapedString;

    use super::*;
    use crate::{
        cli::OutputLogsMode,
        turbo_json::{
            processed::{ProcessedEnv, ProcessedInputs, ProcessedOutputs},
            FutureFlags,
        },
    };

    // Shared test fixtures
    fn create_base_task() -> ProcessedTaskDefinition {
        ProcessedTaskDefinition {
            cache: Some(Spanned::new(true)),
            persistent: Some(Spanned::new(false)),
            outputs: Some(
                ProcessedOutputs::new(
                    vec![Spanned::new(UnescapedString::from("dist/**"))],
                    &FutureFlags::default(),
                )
                .unwrap(),
            ),
            inputs: Some(
                ProcessedInputs::new(
                    vec![Spanned::new(UnescapedString::from("src/**"))],
                    &FutureFlags::default(),
                )
                .unwrap(),
            ),
            env: Some(
                ProcessedEnv::new(
                    vec![Spanned::new(UnescapedString::from("NODE_ENV"))],
                    &FutureFlags::default(),
                )
                .unwrap(),
            ),
            depends_on: None,
            pass_through_env: None,
            output_logs: None,
            interruptible: None,
            interactive: None,
            env_mode: None,
            with: None,
        }
    }

    fn create_override_task() -> ProcessedTaskDefinition {
        ProcessedTaskDefinition {
            cache: Some(Spanned::new(false)),
            persistent: Some(Spanned::new(true)),
            outputs: Some(
                ProcessedOutputs::new(
                    vec![Spanned::new(UnescapedString::from("build/**"))],
                    &FutureFlags::default(),
                )
                .unwrap(),
            ),
            inputs: Some(
                ProcessedInputs::new(
                    vec![Spanned::new(UnescapedString::from("lib/**"))],
                    &FutureFlags::default(),
                )
                .unwrap(),
            ),
            env: Some(
                ProcessedEnv::new(
                    vec![Spanned::new(UnescapedString::from("PROD_ENV"))],
                    &FutureFlags::default(),
                )
                .unwrap(),
            ),
            output_logs: Some(Spanned::new(OutputLogsMode::Full)),
            interruptible: Some(Spanned::new(true)),
            depends_on: None,
            pass_through_env: None,
            interactive: None,
            env_mode: None,
            with: None,
        }
    }

    fn create_partial_task() -> ProcessedTaskDefinition {
        ProcessedTaskDefinition {
            persistent: Some(Spanned::new(true)),
            output_logs: Some(Spanned::new(OutputLogsMode::HashOnly)),
            cache: None,
            outputs: None,
            inputs: None,
            env: None,
            depends_on: None,
            pass_through_env: None,
            interruptible: None,
            interactive: None,
            env_mode: None,
            with: None,
        }
    }

    #[test]
    fn test_other_takes_priority() {
        let mut base = create_base_task();
        let override_task = create_override_task();

        // Store original values for comparison
        let original_cache = base.cache.clone();
        let original_persistent = base.persistent.clone();
        let original_outputs = base.outputs.clone();

        // Perform merge
        base.merge(override_task.clone());

        // All fields from override_task should take priority
        assert_eq!(base.cache, override_task.cache);
        assert_eq!(base.persistent, override_task.persistent);
        assert_eq!(base.outputs, override_task.outputs);
        assert_eq!(base.inputs, override_task.inputs);
        assert_eq!(base.env, override_task.env);
        assert_eq!(base.output_logs, override_task.output_logs);
        assert_eq!(base.interruptible, override_task.interruptible);

        // Verify values actually changed
        assert_ne!(base.cache, original_cache);
        assert_ne!(base.persistent, original_persistent);
        assert_ne!(base.outputs, original_outputs);
    }

    #[test]
    fn test_partial_merge_preserves_existing() {
        let mut base = create_base_task();
        let partial = create_partial_task();

        // Store original values that should be preserved
        let original_cache = base.cache.clone();
        let original_outputs = base.outputs.clone();
        let original_inputs = base.inputs.clone();
        let original_env = base.env.clone();

        // Perform merge
        base.merge(partial.clone());

        // Fields present in partial should be overridden
        assert_eq!(base.persistent, partial.persistent);
        assert_eq!(base.output_logs, partial.output_logs);

        // Fields not present in partial should be preserved
        assert_eq!(base.cache, original_cache);
        assert_eq!(base.outputs, original_outputs);
        assert_eq!(base.inputs, original_inputs);
        assert_eq!(base.env, original_env);

        // Fields not set in either should remain None
        assert_eq!(base.interruptible, None);
        assert_eq!(base.interactive, None);
    }

    #[test]
    fn test_from_iter_last_takes_priority() {
        let first = create_base_task();
        let second = create_partial_task();
        let third = create_override_task();

        let tasks = vec![first.clone(), second.clone(), third.clone()];
        let result: ProcessedTaskDefinition = tasks.into_iter().collect();

        // Fields present in the last task (third) should take priority
        assert_eq!(result.cache, third.cache);
        assert_eq!(result.persistent, third.persistent);
        assert_eq!(result.outputs, third.outputs);
        assert_eq!(result.inputs, third.inputs);
        assert_eq!(result.env, third.env);
        assert_eq!(result.output_logs, third.output_logs);
        assert_eq!(result.interruptible, third.interruptible);

        // Fields only present in earlier tasks should be preserved if not
        // overridden (none in this case since third task overrides all
        // fields that first task had)
    }

    #[test]
    fn test_from_iter_combines_across_multiple_tasks() {
        let first = ProcessedTaskDefinition {
            cache: Some(Spanned::new(true)),
            outputs: Some(
                ProcessedOutputs::new(
                    vec![Spanned::new(UnescapedString::from("dist/**"))],
                    &FutureFlags::default(),
                )
                .unwrap(),
            ),
            ..Default::default()
        };

        let second = ProcessedTaskDefinition {
            persistent: Some(Spanned::new(false)),
            inputs: Some(
                ProcessedInputs::new(
                    vec![Spanned::new(UnescapedString::from("src/**"))],
                    &FutureFlags::default(),
                )
                .unwrap(),
            ),
            ..Default::default()
        };

        let third = ProcessedTaskDefinition {
            env: Some(
                ProcessedEnv::new(
                    vec![Spanned::new(UnescapedString::from("NODE_ENV"))],
                    &FutureFlags::default(),
                )
                .unwrap(),
            ),
            output_logs: Some(Spanned::new(OutputLogsMode::Full)),
            // Override cache from first task
            cache: Some(Spanned::new(false)),
            ..Default::default()
        };

        let tasks = vec![first.clone(), second.clone(), third.clone()];
        let result: ProcessedTaskDefinition = tasks.into_iter().collect();

        // Last task's cache should override first task's cache
        assert_eq!(result.cache, third.cache);

        // Fields from second task should be preserved since not overridden
        assert_eq!(result.persistent, second.persistent);
        assert_eq!(result.inputs, second.inputs);

        // Fields from first task should be preserved since not overridden later
        assert_eq!(result.outputs, first.outputs);

        // Fields from third task
        assert_eq!(result.env, third.env);
        assert_eq!(result.output_logs, third.output_logs);
    }

    #[test]
    fn test_from_iter_empty_iterator() {
        let empty_vec: Vec<ProcessedTaskDefinition> = vec![];
        let result: ProcessedTaskDefinition = empty_vec.into_iter().collect();

        // Should be equivalent to default
        assert_eq!(result, ProcessedTaskDefinition::default());
    }

    #[test]
    fn test_from_iter_single_task() {
        let single_task = create_base_task();
        let tasks = vec![single_task.clone()];
        let result: ProcessedTaskDefinition = tasks.into_iter().collect();

        assert_eq!(result, single_task);
    }

    // Reusable fixtures for array fields
    fn env_base() -> ProcessedEnv {
        ProcessedEnv {
            vars: vec!["BASE_ENV".to_string()],
            extends: false,
        }
    }

    fn env_override() -> ProcessedEnv {
        ProcessedEnv {
            vars: vec!["OVERRIDE_ENV".to_string()],
            extends: false,
        }
    }

    fn env_extending() -> ProcessedEnv {
        ProcessedEnv {
            vars: vec!["OVERRIDE_ENV".to_string()],
            extends: true,
        }
    }

    fn deps_base() -> ProcessedDependsOn {
        ProcessedDependsOn {
            deps: vec![Spanned::new(UnescapedString::from("build"))],
            extends: false,
        }
    }

    fn deps_test() -> ProcessedDependsOn {
        ProcessedDependsOn {
            deps: vec![Spanned::new(UnescapedString::from("test"))],
            extends: false,
        }
    }

    fn deps_extending() -> ProcessedDependsOn {
        ProcessedDependsOn {
            deps: vec![Spanned::new(UnescapedString::from("test"))],
            extends: true,
        }
    }

    fn with_task1() -> ProcessedWith {
        use turborepo_task_id::TaskName;
        ProcessedWith {
            tasks: vec![Spanned::new(TaskName::from("task1"))],
            extends: false,
        }
    }

    fn with_task2_extending() -> ProcessedWith {
        use turborepo_task_id::TaskName;
        ProcessedWith {
            tasks: vec![Spanned::new(TaskName::from("task2"))],
            extends: true,
        }
    }

    fn with_task3() -> ProcessedWith {
        use turborepo_task_id::TaskName;
        ProcessedWith {
            tasks: vec![Spanned::new(TaskName::from("task3"))],
            extends: false,
        }
    }

    fn inputs_base() -> ProcessedInputs {
        ProcessedInputs {
            globs: vec![],
            default: false,
            extends: false,
        }
    }

    fn inputs_extending_with_default() -> ProcessedInputs {
        ProcessedInputs {
            globs: vec![],
            default: true,
            extends: true,
        }
    }

    fn outputs_base() -> ProcessedOutputs {
        ProcessedOutputs {
            globs: vec![],
            extends: false,
        }
    }

    fn outputs_extending() -> ProcessedOutputs {
        ProcessedOutputs {
            globs: vec![],
            extends: true,
        }
    }

    #[test]
    fn test_merge_with_extends_true() {
        let mut base = ProcessedTaskDefinition {
            inputs: Some(inputs_base()),
            outputs: Some(outputs_base()),
            env: Some(env_base()),
            depends_on: Some(deps_base()),
            with: Some(with_task1()),
            ..Default::default()
        };

        let extending = ProcessedTaskDefinition {
            inputs: Some(inputs_extending_with_default()),
            outputs: Some(outputs_extending()),
            env: Some(env_extending()),
            depends_on: Some(deps_extending()),
            with: Some(with_task2_extending()),
            ..Default::default()
        };

        base.merge(extending);

        // Verify extends behavior
        assert!(base.inputs.as_ref().unwrap().default); // OR'd
        assert!(base.inputs.as_ref().unwrap().extends);
        assert_eq!(
            base.env.as_ref().unwrap().vars,
            vec!["BASE_ENV".to_string(), "OVERRIDE_ENV".to_string()]
        );
        assert_eq!(base.depends_on.as_ref().unwrap().deps.len(), 2);
        assert_eq!(base.with.as_ref().unwrap().tasks.len(), 2);
    }

    #[test]
    fn test_merge_with_extends_false() {
        let mut base = ProcessedTaskDefinition {
            env: Some(env_base()),
            depends_on: Some(deps_base()),
            with: Some(with_task1()),
            ..Default::default()
        };

        let replacing = ProcessedTaskDefinition {
            env: Some(env_override()),
            depends_on: Some(deps_test()),
            with: Some(with_task3()),
            ..Default::default()
        };

        base.merge(replacing);

        // Verify replace behavior
        assert_eq!(base.env, Some(env_override()));
        assert_eq!(base.depends_on, Some(deps_test()));
        assert_eq!(base.with, Some(with_task3()));
    }

    #[test]
    fn test_merge_chain_with_extends_then_replace() {
        // Test that when chaining: base -> extending -> replacing
        // The final replacing task overrides everything, not extends

        let base = ProcessedTaskDefinition {
            depends_on: Some(deps_base()), // has "build"
            env: Some(env_base()),         // has "BASE_ENV"
            ..Default::default()
        };

        // Middle task extends the base
        let extending = ProcessedTaskDefinition {
            depends_on: Some(deps_extending()), // has "test" with extends: true
            env: Some(env_extending()),         // has "OVERRIDE_ENV" with extends: true
            ..Default::default()
        };

        // Final task replaces (no extends)
        let replacing = ProcessedTaskDefinition {
            depends_on: Some(ProcessedDependsOn {
                deps: vec![Spanned::new(UnescapedString::from("lint"))],
                extends: false, // This should replace, not extend
            }),
            env: Some(ProcessedEnv {
                vars: vec!["FINAL_ENV".to_string()],
                extends: false, // This should replace, not extend
            }),
            ..Default::default()
        };

        // Apply the chain of merges
        let result = ProcessedTaskDefinition::from_iter(vec![base, extending, replacing]);

        assert_eq!(
            result.depends_on.as_ref().unwrap().deps,
            vec![Spanned::new(UnescapedString::from("lint"))]
        );

        // Verify the extends flags are now false
        assert!(!result.depends_on.as_ref().unwrap().extends);
    }
}
