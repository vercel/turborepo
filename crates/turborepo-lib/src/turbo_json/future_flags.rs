//! Future flags for enabling experimental or upcoming features
//!
//! This module contains the `FutureFlags` structure which allows users to
//! opt-in to experimental features before they become the default behavior.
//!
//! ## Usage
//!
//! Future flags can be configured in the root `turbo.json`:
//!
//! ```json
//! {
//!   "futureFlags": {
//!     "turboExtends": true
//!   }
//! }
//! ```
//!
//! Note: Future flags are only allowed in the root `turbo.json` and will cause
//! an error if specified in workspace packages.

use biome_deserialize_macros::Deserializable;
use serde::Serialize;
use struct_iterable::Iterable;

/// Future flags configuration for experimental features
///
/// Each flag represents an experimental feature that can be enabled
/// before it becomes the default behavior in a future version.
#[derive(Serialize, Default, Debug, Copy, Clone, Iterable, Deserializable, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct FutureFlags {
    /// Enable `$TURBO_EXTENDS$`
    ///
    /// When enabled, allows using `$TURBO_EXTENDS$` in array fields.
    /// This will change the default behavior of overriding the field to instead
    /// append.
    pub turbo_extends: bool,
}

impl FutureFlags {
    /// Create a new FutureFlags with all features disabled
    pub fn new() -> Self {
        Self::default()
    }
}
