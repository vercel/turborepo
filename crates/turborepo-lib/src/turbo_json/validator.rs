use miette::{NamedSource, SourceSpan};
use turborepo_repository::package_graph::{PackageName, ROOT_PKG_NAME};

use super::{Error, TurboJson, TOPOLOGICAL_PIPELINE_DELIMITER};
use crate::config::UnnecessaryPackageTaskSyntaxError;

pub type TurboJSONValidation = fn(&Validator, &TurboJson) -> Vec<Error>;

/// Validator for TurboJson structures with context-aware validation
pub struct Validator;

const ROOT_VALIDATIONS: &[TurboJSONValidation] = &[validate_with_has_no_topo];
const PACKAGE_VALIDATIONS: &[TurboJSONValidation] = &[
    validate_with_has_no_topo,
    validate_no_package_task_syntax,
    validate_extends,
];

impl Validator {
    /// Creates a new validator instance
    pub fn new() -> Self {
        Self
    }

    /// Validates a TurboJson based on its package context
    ///
    /// Root turbo.json files have different validation rules than workspace
    /// turbo.json files
    pub fn validate_turbo_json(
        &self,
        package_name: &PackageName,
        turbo_json: &TurboJson,
    ) -> Vec<Error> {
        let validations = match package_name {
            PackageName::Root => ROOT_VALIDATIONS,
            PackageName::Other(_) => PACKAGE_VALIDATIONS,
        };
        validations
            .iter()
            .flat_map(|validation| validation(self, turbo_json))
            .collect()
    }
}

impl Default for Validator {
    fn default() -> Self {
        Self::new()
    }
}

pub fn validate_no_package_task_syntax(
    _validator: &Validator,
    turbo_json: &TurboJson,
) -> Vec<Error> {
    turbo_json
        .tasks
        .iter()
        .filter(|(task_name, _)| task_name.is_package_task())
        .map(|(task_name, entry)| {
            let (span, text) = entry.span_and_text("turbo.json");
            Error::UnnecessaryPackageTaskSyntax(Box::new(UnnecessaryPackageTaskSyntaxError {
                actual: task_name.to_string(),
                wanted: task_name.task().to_string(),
                span,
                text,
            }))
        })
        .collect()
}

pub fn validate_extends(_validator: &Validator, turbo_json: &TurboJson) -> Vec<Error> {
    match turbo_json.extends.first() {
        Some(package_name) if package_name != ROOT_PKG_NAME || turbo_json.extends.len() > 1 => {
            let (span, text) = turbo_json.extends.span_and_text("turbo.json");
            vec![Error::ExtendFromNonRoot { span, text }]
        }
        None => {
            let path = turbo_json
                .path
                .as_ref()
                .map_or("turbo.json", |p| p.as_ref());

            let (span, text) = match turbo_json.text {
                Some(ref text) => {
                    let len = text.len();
                    let span: SourceSpan = (0, len - 1).into();
                    (Some(span), text.to_string())
                }
                None => (None, String::new()),
            };

            vec![Error::NoExtends {
                span,
                text: NamedSource::new(path, text),
            }]
        }
        _ => vec![],
    }
}

pub fn validate_with_has_no_topo(_validator: &Validator, turbo_json: &TurboJson) -> Vec<Error> {
    turbo_json
        .tasks
        .iter()
        .flat_map(|(_, definition)| {
            definition.with.iter().flatten().filter_map(|with_task| {
                if with_task.starts_with(TOPOLOGICAL_PIPELINE_DELIMITER) {
                    let (span, text) = with_task.span_and_text("turbo.json");
                    Some(Error::InvalidTaskWith { span, text })
                } else {
                    None
                }
            })
        })
        .collect()
}

#[cfg(test)]
mod test {
    use std::assert_matches::assert_matches;

    use test_case::test_case;
    use turborepo_errors::Spanned;
    use turborepo_task_id::TaskName;
    use turborepo_unescape::UnescapedString;

    use super::*;
    use crate::turbo_json::{Pipeline, RawTaskDefinition};

    #[test]
    fn test_validate_with_has_no_topo() {
        let turbo_json = TurboJson {
            tasks: Pipeline(
                vec![(
                    TaskName::from("dev"),
                    Spanned::new(RawTaskDefinition {
                        with: Some(vec![Spanned::new(UnescapedString::from("^proxy"))]),
                        ..Default::default()
                    }),
                )]
                .into_iter()
                .collect(),
            ),
            ..Default::default()
        };

        let validator = Validator::new();
        let errs = validate_with_has_no_topo(&validator, &turbo_json);
        let error_messages: Vec<String> = errs.iter().map(|e| e.to_string()).collect();
        insta::assert_debug_snapshot!("validate_with_has_no_topo", error_messages);
    }

    #[test_case(
        vec!["my-package#build"],
        "single_package_task"
    )]
    #[test_case(
        vec!["build"],
        "non_package_task"
    )]
    #[test_case(
        vec!["pkg-a#test", "pkg-b#lint", "build"],
        "multiple_mixed_tasks"
    )]
    fn test_validate_no_package_task_syntax(tasks: Vec<&str>, name: &str) {
        let turbo_json = TurboJson {
            tasks: Pipeline(
                tasks
                    .into_iter()
                    .map(|task_name| {
                        (
                            TaskName::from(task_name.to_string()),
                            Spanned::new(RawTaskDefinition::default()),
                        )
                    })
                    .collect(),
            ),
            ..Default::default()
        };

        let validator = Validator::new();
        let errs = validate_no_package_task_syntax(&validator, &turbo_json);
        let error_messages: Vec<String> = errs.iter().map(|e| e.to_string()).collect();
        let snapshot_name = format!("validate_no_package_task_syntax_{}", name);
        insta::assert_debug_snapshot!(snapshot_name, error_messages);
    }

    #[test_case(
        vec![],
        "no_extends"
    )]
    #[test_case(
        vec!["//"],
        "valid_extends_from_root"
    )]
    #[test_case(
        vec!["some-package"],
        "extends_from_non_root_package"
    )]
    #[test_case(
        vec!["//", "other-package"],
        "multiple_extends_including_root"
    )]
    #[test_case(
        vec!["package-a", "package-b"],
        "multiple_extends_not_including_root"
    )]
    fn test_validate_extends(extends: Vec<&str>, name: &str) {
        let turbo_json = TurboJson {
            extends: Spanned::new(extends.into_iter().map(String::from).collect()),
            ..Default::default()
        };

        let validator = Validator::new();
        let errs = validate_extends(&validator, &turbo_json);
        let error_messages: Vec<String> = errs.iter().map(|e| e.to_string()).collect();
        let snapshot_name = format!("validate_extends_{}", name);
        insta::assert_debug_snapshot!(snapshot_name, error_messages);
    }

    #[test]
    fn test_validator_with_root_package() {
        let validator = Validator::new();

        // Root turbo.json can have package task syntax
        let turbo_json = TurboJson {
            tasks: Pipeline(
                vec![(TaskName::from("app#build"), Spanned::default())]
                    .into_iter()
                    .collect(),
            ),
            ..Default::default()
        };

        let errs = validator.validate_turbo_json(&PackageName::Root, &turbo_json);
        assert!(
            errs.is_empty(),
            "Root turbo.json should allow package task syntax"
        );
    }

    #[test]
    fn test_validator_with_missing_extends() {
        let validator = Validator::new();

        // Workspace turbo.json without extends should error
        let turbo_json = TurboJson {
            ..Default::default()
        };

        let errs = validator.validate_turbo_json(&PackageName::from("app"), &turbo_json);
        assert_eq!(errs.len(), 1, "Workspace turbo.json should have extends");
        assert_matches!(errs[0], Error::NoExtends { .. });
    }
}
