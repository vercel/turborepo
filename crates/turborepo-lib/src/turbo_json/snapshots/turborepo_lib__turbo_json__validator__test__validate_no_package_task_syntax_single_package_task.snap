---
source: crates/turborepo-lib/src/turbo_json/validator.rs
expression: error_messages
---
[
    "\"my-package#build\". Use \"build\" instead.",
]
